osbs/build/user_params_meta.py

```diff
@@ -0,0 +1,176 @@
+"""
+Copyright (c) 2020 Red Hat, Inc
+All rights reserved.
+
+This software may be modified and distributed under the terms
+of the BSD license. See the LICENSE file for details.
+"""
+
+from __future__ import absolute_import, unicode_literals
+
+import six
+
+from osbs.exceptions import OsbsValidationException
+
+
+class BuildParam(object):
+    """
+    One parameter of a spec.
+
+    Works like a data descriptor (similar to `property`), should be defined as
+    a class attribute on classes that inherit from BuildParamsBase.
+
+    When defined as a class attribute, the name of the attribute should match
+    the name given to the BuildParam in __init__(), e.g.:
+
+    >>> class BuildParams(BuildParamsBase)
+    >>>     x = BuildParam("x")
+    >>>     y = BuildParam("y")
+
+    BuildParamsBase and classes that inherit from it will validate this during
+    class creation.
+    """
+
+    def __init__(self, name, default=None, required=False, include_in_json=True):
+        """
+        Define a BuildParam.
+
+        Do NOT use mutable values as default, same issues apply as when using
+        mutable defaults for method params.
+
+        :param name: str, name of param, should match name of class attribute
+        :param default: any immutable object, default value of param
+        :param required: bool, is the param allowed to be None?
+        :param include_in_json: bool, include the param in to_json() output?
+        """
+        self._name = name
+        self._default = default
+        self._required = required
+        self._include_in_json = include_in_json
+        self._mangled_name = "_{self.__class__.__name__}__{self.name}".format(self=self)
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def required(self):
+        return self._required
+
+    @property
+    def include_in_json(self):
+        return self._include_in_json
```

Comment:
I'm paranoid so I often use properties to prevent others (and myself) from changing attributes that should be immutable. There is also the possibility that some subclass may want to override the `include_in_json` (or, more likely, `required`) property. But I'm not strictly opposed to making this a simple attribute, let's see what the others think.

Reply:
Will keep it as it is for now